### ISSUE TITLE
Float values pulled from Odoo are displayed as Integer

### ISSUE THREAD

**Issue by livingdeadfr** [tbondois/odoo-ripcord#5]
Hi M. Bondois

I use the 1.7 version of your library.

When I query a list of amount, they all arrive in php as rounded number (type: float) (ex odoo 24.50 => 24 in php).

The original datas in odoo are decimal (float), and are displayed correctly in odoo.

I track up to the : `public function read(string $model, array $ids, array $fields = [], array $context = [])`
And yet the value is rounded in the response.

Is there any issue regarding execute_kw ? DarkOnline ripcord is under version 1.8, is this linked ? 

Best regards

David

**Comment by tbondois** [comment]
Hi, David,

Thanks for reporting that issue ! I'm not aware of it, and as I remember there is not type conversion on my code.

But it's based on darkaonline/ripcord, and I never checked the updated he did since that. But the goal of theses libraries is too be modular, so that will be very curious

First, 2 things to check :

1) Can you check what return the odoo api, as a raw response ? You can do that using tools like curl, Postman, or var_dump() the raw response in my `read()` function (ie: `var_dump($this->getModelService()->_getRawResponse());` or something like that)

2) Trying downgrading to the 0.1.6 version of darkaonline/ripcord : 
`composer require tbondois/odoo-ripcord darkaonline/ripcord:0.1.6`

I don't work with Odoo anymore, but I will try to help. If you find how to fix that problem, don't hesitate to Open a Pull Request on this repository. 

But I don't really have time to fix bugs in it by myself. So since darkaonline/ripcord is still maintained, maybe you should considering using his library instead of mine, for long-term use.

**Comment by livingdeadfr** [comment]
Hello Thomas,

On the following call :

>
>     public function read(string $model, array $ids, array $fields = [],
> array $context = [])
>     {
>         $response = $this->getModelService()->execute_kw(
>             $this->db, $this->uid(), $this->password,
>             $model,
>             'read',
>             [$ids],
>             ['fields' => $fields, 'context' => $context]
>         );
>         return $this->setResponse($response);

With those paraméters :
$model : "cap_deadline"
$ids : [1436,1373,1295,...]
$fields :  array[8]
[0] string "status"
[1] string "id"
[2] string "due_date"
[3] string "invoice_number"
[4] string "expected_amount"
[5] string "timetable_id"
[6] string "payment_ids"
[7] string "validation_date"

$context : [] (empty)

 $response :
[0] array[8]
[due_date] string "2022-12-05"
*[expected_amount] float 24 *
[id] integer 1436
[invoice_number] string "FCC/2022/0005"
[payment_ids] array[0]
[status] string "waiting"
[timetable_id] array[2]
[validation_date] boolean false

In the database (postgresql) accessible from Odoo, the value is 24,55, as
we can see in odoo

[image: image.png]

Do you have any idea, to drive my search ?

Kind regards

David

Le mar. 22 févr. 2022 à 19:16, Thomas Bondois ***@***.***> a
écrit :

**Comment by tbondois** [comment]
I can see It's not a type conversion since you have a `float 24`, not `integer 24`. It's not a `round()`, else 24.55 will turn into 25.
a `floor()` could do that, as its return a float after rounding down, but there is not such things in my library.

There is something I think in one your setup rounding either :
I) rounding down floats (very unlikely)
II) or badly converting (like the comma `,` from odoo response is not converted as a decimal `.` in PHP). (I will bet more on that, but I dunno how that could happens)

It Can come from :
a) your project vendors (composer libraries) 
b) or server (PHP config, PHP extensions, fake PHP extensions as composer libraries) 


To check if it's not an issue with the server :
1. Do a `var_dump(24.55);` at the same place you did the previous `var_dump`. Is it rounded down on display ?

To check if it's not from from a library of your project :
2. Search the keyword `floor` in your `vendor/` . Does it appear in darkaonline/ripcord lib ?
3. Check if it happens again with a `composer require tbondois/odoo-ripcord darkaonline/ripcord:0.1.6` (to downgrade ripcord to a version I used when i was developping mine)
4. If the problem still here, dig into darkaonline/ripcord to debug how if manage float data types (and if it do operations on comma or dot)
5. You don't find nothing ? search in the github issues (open & closed) of darkaonline/ripcord if that problem was already reported ; if not, open one

**Comment by livingdeadfr** [comment]
Hi Thomas,

I did found the mistake.

I discovered that the XML file got from odoo was correct in getService but
got wrong after conversation to PHP XML object.

My program is on a French server with the locale set to fr_FR.UTF-8. While
the conversation in object occurs it convert double to floats. But the XML
value is in US format. Whereas the Fr decimal separator is "," . so the
value is trunk.

I have to switch to US format and set back to French after it.

Thanks for your follow up. And hope it could help others.

Kind regards

David

Le mar. 1 mars 2022, 18:13, Thomas Bondois ***@***.***> a
écrit :
